### PR TITLE
implement new interface to unify session authentication in client

### PR DESCRIPTION
**Pull request by FedericoAmura**
# Description

This PR changes the way PKP packages handle `sessionSigs`. Now instead of receiving it directly they will obtain it from the client and its auth context. This way we are unifying that management and using them from a common storage instead of making the user handle it directly

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit test to validate PKP wallet proper creation
- [x] e2e test to sign using the sessionSigs from the common storage

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/pkp-base/src/lib/pkp-base.ts [hunk 1]
```diff
@@ -222,6 +200,14 @@ export class PKPBase<T = PKPBaseDefaultParams> {
     }
   }
 
+  private validateAuthContext() {
+    if (this.controllerAuthSig && this.authContext) {
+      throw new Error(
```
Comment (Ansonhkg) [inline packages/pkp-base/src/lib/pkp-base.ts:205]: can we use `this.throwError` in this class instead? we probably need to clean up in other places as well in this file

File: packages/pkp-base/src/lib/pkp-base.ts [hunk 2]
```diff
@@ -243,26 +228,29 @@ export class PKPBase<T = PKPBaseDefaultParams> {
...
-    }
+    this.validateAuthContext();
 
-    // If session sigs are provided, they must be an object
+    // If auth context is provided, it must have a client, props and auth methods
     if (
-      this.controllerSessionSigs &&
-      typeof this.controllerSessionSigs !== 'object'
+      this.authContext &&
+      typeof this.authContext?.client !== 'object' &&
```
Comment (Ansonhkg) [inline packages/pkp-base/src/lib/pkp-base.ts:236]: `typeof` null will also be an object so we gotta check `this.authContext.client !== null` too I think 

image

File: packages/pkp-base/src/lib/pkp-base.ts [hunk 3]
```diff
@@ -324,11 +321,12 @@ export class PKPBase<T = PKPBaseDefaultParams> {
       throw new Error('pkpPubKey (aka. uncompressPubKey) is required');
     }
 
-    if (this.controllerAuthSig && this.controllerSessionSigs) {
-      throw new Error(
-        'controllerAuthSig, controllerSessionSigs and controllerAuthMethod are defined, can only use one authorization type'
+    this.validateAuthContext();
+
```
Comment (Ansonhkg) [inline packages/pkp-base/src/lib/pkp-base.ts:325]: shouldn't we have this code as well here?
```js
if (
  this.authContext &&
  typeof this.authContext?.client !== 'object' &&
  typeof this.authContext?.getSessionSigsProps !== 'function' &&
  !Array.isArray(this.authContext?.authMethods)
) {
  throw new Error('controllerSessionSigs must be an object');
}
```

**Review by Ansonhkg — CHANGES_REQUESTED** [review]
We no longer accept `controllerSessionSigs`, `controllerAuthMethods` and `sessionSigsExpiration` as params, so i think is a breaking change, requires docs update and a migration guide. Also we need to update the e2e tests too if we are using these fields eg. `test-pkp-ethers-handler-sign-tx.mjs`

I think this should be a major bump

**Review by joshLong145 — COMMENTED** [review]
looks good just one more additional comments

File: packages/pkp-base/src/lib/pkp-base.ts (hunk 1 above)
Comment (FedericoAmura) [reply packages/pkp-base/src/lib/pkp-base.ts]: Changed

File: packages/pkp-base/src/lib/pkp-base.ts (hunk 2 above)
Comment (FedericoAmura) [reply packages/pkp-base/src/lib/pkp-base.ts]: Changed

File: packages/pkp-base/src/lib/pkp-base.ts (hunk 3 above)
Comment (FedericoAmura) [reply packages/pkp-base/src/lib/pkp-base.ts:330]: You are right. Added

File: e2e-nodejs/group-pkp-session-sigs/test-pkp-sign-with-session-sigs-session-cache.mjs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    pkpPubKey: globalThis.LitCI.AUTH_METHOD_PKP_INFO.publicKey,
+    rpc: LITCONFIG.CHRONICLE_RPC,
+    litNetwork: globalThis.LitCI.network,
+    authContext: {
+      client,
+      getSessionSigsProps: {
+        chain: 'ethereum',
+        expiration: new Date(Date.now() + 60_000 * 60).toISOString(),
+        resourceAbilityRequests: resourceAbilities,
+        sessionKey: sessionKeyPair,
```
[the remark this answers is not in the document]
Comment (FedericoAmura) [reply e2e-nodejs/group-pkp-session-sigs/test-pkp-sign-with-session-sigs-session-cache.mjs]: It was loaded from storage. But it was obtained anyway when getting sessionSigs so I removed it to make that option explicit

**Comment by FedericoAmura** [comment]
> We no longer accept `controllerSessionSigs`, `controllerAuthMethods` and `sessionSigsExpiration` as params, so i think is a breaking change, requires docs update and a migration guide. Also we need to update the e2e tests too if we are using these fields eg. `test-pkp-ethers-handler-sign-tx.mjs`
> 
> I think this should be a major bump

We are backwards compatible now =D

**Review by Ansonhkg — COMMENTED** [review]
Niccee!! 

Can we also update the docs? I have created a branch here https://github.com/LIT-Protocol/docs/pull/226 you can continue what I have edited the file. It should reflect to this page:
https://developer.litprotocol.com/v3/sdk/wallets/walletconnect/#1-create-a-pkpclient

**Comment by FedericoAmura** [comment]
> Niccee!!
> 
> Can we also update the docs? I have created a branch here [LIT-Protocol/docs#226](https://github.com/LIT-Protocol/docs/pull/226) you can continue what I have left there.

Updated!

File: packages/types/src/lib/interfaces.ts [hunk 5]
```diff
@@ -1094,15 +1094,44 @@ export declare type AuthenticatorAttachment = 'cross-platform' | 'platform';
...
+    sessionKeyUri: any;
+    resourceAbilityRequests: Array<LitResourceAbilityRequest>;
+  }) => Promise<boolean>;
+  getSessionSigs: (params: GetSessionSigsProps) => Promise<SessionSigsMap>;
+  signSessionKey: (
+    params: SignSessionKeyProp
+  ) => Promise<SignSessionKeyResponse>;
+}
+
+export interface AuthenticationProps {
```
Comment (joshLong145) [inline packages/types/src/lib/interfaces.ts:1117]: this looks good but i think we also need `authNeededCallback` for if they wish to override the default implementation with out and explicit call to `getSessionSigs`

**Review by joshLong145 — COMMENTED** [review]
Overall looks good, left a comment on `AuthenticationProps`

**Pull request merged**